data/transactions/logic/assembler.go

```diff
@@ -1182,13 +1214,354 @@ const AssemblerDefaultVersion = 1
+	type byteFrequency struct {
+		value []byte
+		freq  int
+	}
+
+	if ops.hasBytecBlock {
+		// don't optimize an existing bytecblock, only byte/addr pseudo-ops
+		return nil
+	}
+
+	freqs := make([]byteFrequency, len(ops.bytec))
+
+	for i, value := range ops.bytec {
+		freqs[i].value = value
+	}
+
+	for _, ref := range ops.bytecRefs {
+		found := false
+		for i := range freqs {
+			if bytes.Compare(ref.value, freqs[i].value) == 0 {
+				freqs[i].freq++
+				found = true
+				break
+			}
+		}
+		if !found {
+			return ops.errorf("Value not found in bytec block: %v", ref.value)
+		}
+	}
+
+	for _, f := range freqs {
+		if f.freq == 0 {
+			return ops.errorf("Member of bytec block is not used: %v", f.value)
+		}
+	}
+
+	// sort values by greatest to smallest frequency
+	// since we're using a stable sort, constants with the same frequency
+	// will retain their current ordering (i.e. first referenced, first in byte block)
+	sort.SliceStable(freqs, func(i, j int) bool {
+		return freqs[i].freq > freqs[j].freq
+	})
+
+	// sort refs from last to first in raw
+	// this way when we iterate through them and potentially change the size of raw,
+	// the later positions will not affect the indexes of the earlier positions
+	sort.Slice(ops.bytecRefs, func(i, j int) bool {
+		return ops.bytecRefs[i].position > ops.bytecRefs[j].position
+	})
+
+	var scratch [binary.MaxVarintLen64]byte
+	raw := ops.pending.Bytes()
+	for _, ref := range ops.bytecRefs {
+		singleton := false
+		newIndex := -1
+		for i, f := range freqs {
+			if bytes.Compare(ref.value, f.value) == 0 {
+				singleton = f.freq == 1
+				newIndex = i
+				break
+			}
+		}
+		if newIndex == -1 {
+			return ops.errorf("Value not found in bytec block: %v", ref.value)
+		}
+
+		var newBytes []byte
+		if singleton {
+			vlen := binary.PutUvarint(scratch[:], uint64(len(ref.value)))
+			newBytes = make([]byte, 1+vlen+len(ref.value))
+			newBytes[0] = 0x80 // pushbytes
+			copy(newBytes[1:], scratch[:vlen])
+			copy(newBytes[1+vlen:], ref.value)
+		} else {
+			switch newIndex {
+			case 0:
+				newBytes = []byte{0x28} // bytec_0
+			case 1:
+				newBytes = []byte{0x29} // bytec_1
+			case 2:
+				newBytes = []byte{0x2a} // bytec_2
+			case 3:
+				newBytes = []byte{0x2b} // bytec_3
+			default:
+				newBytes = []byte{0x27, uint8(newIndex)} // bytec {index}
+			}
+		}
+
+		var currentBytesLen int
+		switch raw[ref.position] {
+		case 0x28:
+			fallthrough
+		case 0x29:
+			fallthrough
+		case 0x2a:
+			fallthrough
+		case 0x2b:
+			currentBytesLen = 1
+		case 0x27:
+			currentBytesLen = 2
+		default:
+			return ops.errorf("Unexpected op at byteReference: %d, from source line %d", raw[ref.position], ops.OffsetToLine[ref.position])
+		}
+
+		raw = replaceBytes(raw, ref.position, currentBytesLen, newBytes)
+		positionDelta := len(newBytes) - currentBytesLen
+
+		for i := range ops.intcRefs {
+			if ops.intcRefs[i].position > ref.position {
+				ops.intcRefs[i].position += positionDelta
+			}
+		}
```

Comment:
I'm a little confused why you have to change the refs of ints in while optimizing bytes AND vice versa. I would think you'd only have to do the one that happens second.  Maybe you're doing it for symmetry and the refs won't actually be used again?  Could you comment a bit about this movement?

I am wondering if you'd be better off with one big rewrite, from pending into a new, longer byte buffer, rather than sliding things around for each constant.

Reply:
Technically this correction is only necessary in the `optimizeIntcBlock` function, since it runs first and must update the byte references. However it seemed weird to enforce and document that ints must be optimized before bytes, so I added the equivalent logic to `optimizeBytecBlock`, even though it's practically a no-op as is. I'll also add a comment for this.

Reply:
> I am wondering if you'd be better off with one big rewrite, from pending into a new, shorter byte buffer, rather than sliding things around for each constant.

Perhaps that would be better for performance, though it wouldn't simplify any logic. If the performance impact of a change like this were to be significant, I can make a separate PR for it.